### DISCUSSION TITLE
Moving Api routers into Individual files

### DISCUSSION THREAD

**Discussion by Butch78** [SeaQL/sea-orm#1336] [Q&A] [answered]
Hi, 

I was trying to modify this [example](https://github.com/SeaQL/sea-orm/blob/master/examples/axum_example/api/src/lib.rs) to add more entities and it quickly become confusing with multiple entities, what is the best practice for creating routes for multiple entities and moving them into seperate files, I tried myself but I became confused by how to correctly implement the with_state() function. 

As shown below I should be able to change the following:

```rust 
let app = Router::new()
        .route("/", get(list_posts).post(create_post))
        .route("/:id", get(edit_post).post(update_post))
        .route("/new", get(new_post))
        .route("/delete/:id", post(delete_post))
        .nest_service(
            "/static",
            get_service(ServeDir::new(concat!(
                env!("CARGO_MANIFEST_DIR"),
                "/static"
            )))
            .handle_error(|error: std::io::Error| async move {
                (
                    StatusCode::INTERNAL_SERVER_ERROR,
                    format!("Unhandled internal error: {}", error),
                )
            }),
        )
        .layer(CookieManagerLayer::new())
        .with_state(state);
        
```

into something simpler like this:

```rust 

mod post;
mod people;
mod houses;


let app = Router::new()
        .nest("/people", people::people_routes)
        .nest("/house", house::house_routes)
        .next("/post", post::post_routes)
        .nest_service(
            "/static",
            get_service(ServeDir::new(concat!(
                env!("CARGO_MANIFEST_DIR"),
                "/static"
            )))
            .handle_error(|error: std::io::Error| async move {
                (
                    StatusCode::INTERNAL_SERVER_ERROR,
                    format!("Unhandled internal error: {}", error),
                )
            }),
        )
        .layer(CookieManagerLayer::new())
        .with_state(state);
        
```

/post.rs

```rust

pub fn post_routes(state: AppState) -> Router<AppState> {
    Router::new()
        .route("/", get(list_posts).post(create_post))
        .with_state(state)
        .route("/:id", get(edit_post).post(update_post))
        .with_state(state)
        .route("/new", get(new_post))
        .with_state(state)
        .route("/delete/:id", post(delete_post))
        .with_state(state)
}
```

But I am confused how to implement **watch_state()** correctly

**Comment by Butch78** [comment, marked as answer]
I was able to figure it out using this example :) 

https://github.com/brooks-builds/full-stack-todo-rust-course

**Comment by billy1624** [comment]
Hey @Butch78, thanks for the sharing!! Would you like to submit a PR to restructure the example and perhaps add a few more useful API endpoints?

**Reply by Butch78** [reply]
Yeah of course, I am busy studying for my exam until the 16th of Jan but after that, I'll submit a pull request :) 

It's a great stater-pack as @tyt2y3 mentioned and I would love to hear what you think about my small changes to make it more modular

**Comment by tyt2y3** [comment]
I think the examples are meant to be just 'quick demo', but it grew to become a starter-pack over time. So I'd say more structure is good.
